logs/querysample/tracing.go

```diff
@@ -0,0 +1,152 @@
+package querysample
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"log"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/pganalyze/collector/state"
+	"github.com/pganalyze/collector/util"
+	pg_query "github.com/pganalyze/pg_query_go/v4"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+func urlToSample(server *state.Server, grant state.GrantLogs, sample state.PostgresQuerySample) string {
+	fp := util.FingerprintQuery(sample.Query, server.Config.FilterQueryText, -1)
+	fpBin := make([]byte, 8)
+	binary.BigEndian.PutUint64(fpBin, fp)
+
+	return fmt.Sprintf(
+		"%s/databases/%s/queries/%s/samples/%d?role=%s",
+		grant.Config.ServerURL,
+		sample.Database,
+		hex.EncodeToString(fpBin),
+		sample.OccurredAt.Unix(),
+		sample.Username,
+	)
+}
+
+func parseTags(query string) map[string]string {
+	tokens, err := pg_query.Scan(query)
+	if err != nil {
+		return nil
+	}
+
+	tags := make(map[string]string)
+	for _, token := range tokens.GetTokens() {
+		var comment string
+		if token.Token == pg_query.Token_SQL_COMMENT {
+			comment = query[token.Start+2 : token.End]
+		} else if token.Token == pg_query.Token_C_COMMENT {
+			comment = query[token.Start+2 : token.End-2]
+		} else {
+			continue
+		}
+		for _, part := range strings.Split(strings.TrimSpace(comment), ",") {
+			if strings.Contains(part, "=") {
+				keyAndValue := strings.SplitN(part, "=", 2)
+				tags[keyAndValue[0]] = keyAndValue[1]
+				// TODO: Handle sqlcommenter's URL encoding logic, and quoting with single quotes
+			} else if strings.Contains(part, ":") {
+				keyAndValue := strings.SplitN(part, ":", 2)
+				tags[keyAndValue[0]] = keyAndValue[1]
+			} // TODO: Support other formats
+		}
+	}
+	return tags
+}
+
+func initProvider(ctx context.Context, endpoint string) (*sdktrace.TracerProvider, func(context.Context) error, error) {
+	res, err := resource.New(ctx,
```

Comment:
It might be nice to alias this import so we're not left wondering what this super-generic "resource" package is here when reading this later.